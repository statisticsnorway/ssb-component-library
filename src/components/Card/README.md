Card
========

> Wrapper for a Profile Box

### Usage

#### HTML

```html
<div class="ssb-card">
  <button class="clickable" onclick="do something">
    <div class="card-content"> <!-- add class 'centered' to center all content -->
      {fill with content}
    </div>
  </button>
  /* Optional download field */
  <a download href=" " class="download-section">
    {feather.download}
    <span>Last ned</span>
  </a>
</div>

<!-- With image on left side -->
<div class="ssb-card">
  <button class="clickable left-orientation" onclick="do something">
    <div class="card-content with-image">
      {fill with content}
    </div>
  </button>
</div>

<!-- With image on top side -->
<div class="ssb-card">
  <button class="clickable top-orientation" onclick="do something">
    <div class="card-content with-image">
      {fill with content}
    </div>
  </button>
</div>
```

#### React

```jsx harmony
<Card centered fileLocation=" " onClick={() => {}}>
  fill with content
</Card>
```

Available props:

| Name       | Type           | Description  |
| ---------- | ------------- | ----- |
| centered | bool | Centers text |
| children | node | Required. Fills box with content |
| fileLocation | string | Path to downloadable file. Ads the download field |
| image | element | Image element |
| imageElement | oneOf(['left', 'top']) | Position of the image |
| onClick | func | Callback function |
