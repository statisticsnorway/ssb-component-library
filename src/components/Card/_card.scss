.ssb-card {
  @include roboto;

  .clickable {
    @include focus-marker;
    @include reset-button;

    .card-image {
      display: flex;

      img {
        height: 214px;
        width: 380px;
      }
    }

    &.left-orientation {
      display: inline-flex;
      flex-direction: row;
      width: 100%;

      .card-content {
        height: 214px;
        width: 100%;
      }

      @media #{$mobile} {
        flex-direction: column;

        .card-content {
          height: 100%;
          max-width: 380px;
        }

        .card-image {
          width: 100%;
        }

        img  {
          max-width: 380px;
          width: 100%;
        }
      }
    }

    &.top-orientation {
      display: inline-flex;
      flex-direction: column;

      .card-content {
        max-width: 380px;
        width: 100%;
      }

      @media #{$mobile} {
        width: 100%;

        .card-image {
          width: 100%;
        }

        img  {
          max-width: 380px;
          width: 100%;
        }

      }
    }

    &:hover {
      text-decoration: none;

      &.left-orientation {

        .card-content {
          border-left: 5px solid $ssb-green-4;
          padding-left: 16px;
          @media #{$mobile} {
            border-left: 1px solid $ssb-green-4;
            border-top: 5px solid $ssb-green-4;
            padding-left: 20px;
            padding-top: 15px;

            &.with-image {
              padding-top: 16px;
            }
          }
        }
      }

      &.top-orientation {

        .card-content {
          border-top: 5px solid $ssb-green-4;

          &.with-image {
            padding-top: 16px;
          }
        }
      }

      .card-content {
        border: 1px solid $ssb-green-4;

        &:not(.with-image) {
        border-top: 5px solid $ssb-green-4;
        }

        &.with-image {

          .card-title {
            background-color: $ssb-green-4;
            border-bottom: none;
            color: $ssb-white;
            margin-bottom: 10px;
          }
        }

        .arrow-icon {
          border: 2px solid $ssb-green-4;
          border-radius: 50%;
          margin-bottom: 1px;
          margin-top: 13px;
          margin-left: 1px;
          padding: 1px;
        }

        .card-action {
          align-items: center;
          background-color: $ssb-green-4;
          display: flex;
          width: fit-content;

          .arrow-icon {
            border: none;
            color: $ssb-white;
            margin-top: 0;
            padding: 0;
          }

          .href-text {
            color: $ssb-white;
          }
        }

      }
    }
  }

  .card-content {
    background: $ssb-white;
    border: 1px solid $ssb-dark-2;
    border-top: 5px solid $ssb-green-3;
    box-sizing: border-box;
    cursor: pointer;
    display: inline-block;
    flex-direction: column;
    padding: 20px;
    width: 100%;

    &.with-image {
      border-top: 1px solid $ssb-dark-2;

      .card-title {
        border-bottom: 2px solid $ssb-green-4;
        display: table;
      }
    }

    &.profiled {
      align-items: center;
      display: inline-flex;
      text-align: center;

      .card-title {
        font-size: 28px;
      }

      .card-icon {
        margin-bottom: 20px;
        max-height: 132px;
        max-width: 132px;
        width: 132px;
      }
    }

    .arrow-icon {
      box-sizing: initial;
      color: $ssb-green-4;
      display: block;
      margin-top: 12px;
      padding: 4px;
    }

    .card-action {
      @include open-sans;
      align-items: center;
      display: flex;
      margin-top: 12px;

      .arrow-icon {
        margin-top: 0;
        padding: 0;
      }

      .href-text {
        color: $ssb-dark-6;
        font-size: 16px;
        font-weight: bold;
        padding: 4px;
      }
    }

    .card-icon {
      margin-bottom: 16px;
      max-height: 28px;
      max-width: 28px;
      width: 28px;
    }

    .card-subtitle {
      @include open-sans;
      font-size: 14px;
      margin-bottom: 8px;
    }

    .card-title {
      display: block;
      font-size: 20px;
      font-weight: bold;
      margin-bottom: 8px;
    }
  }

  .download-section {
    @include open-sans;
    align-items: center;
    border: 1px solid $ssb-dark-2;
    cursor: pointer;
    display: flex;
    height: 60px;
    max-width: 380px;
    text-decoration: none;

    &:focus {
      @include focus-ring;
    }

    &:hover {
      border: 1px solid $ssb-green-3;
    }

    .download-icon {
      color: $ssb-green-4;
      margin: 0 18px;
    }

    span {
      color: $ssb-dark-6;
      font-size: 16px;
    }
  }
}
