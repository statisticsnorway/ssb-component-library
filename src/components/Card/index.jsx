import React from 'react';
import PropTypes from 'prop-types';
import { Download } from 'react-feather';

const Card = ({
	centered, children, downloadText, fileLocation, image, imagePlacement, onClick,
}) => (
	<div className="ssb-card">
		<button
			className={`clickable ${imagePlacement === 'left' ? 'left-orientation' : 'top-orientation'}`}
			onClick={onClick}
		>
			{image}
			<div className={`card-content${centered ? ' centered' : ''}${image ? ' with-image' : ''}`}>

				{ children }
			</div>
		</button>
		{fileLocation && (
			<a download href={fileLocation} className="download-section">
				<Download className="download-icon" />
				<span>{downloadText}</span>
			</a>
		)}
	</div>
);

Card.defaultProps = {
	centered: false,
	downloadText: 'Last ned',
	imagePlacement: 'top',
	onClick: () => {},
};

Card.propTypes = {
	centered: PropTypes.bool,
	children: PropTypes.node.isRequired,
	downloadText: PropTypes.string,
	fileLocation: PropTypes.string,
	image: PropTypes.element,
	imagePlacement: PropTypes.oneOf(['left', 'top']),
	onClick: PropTypes.func,
};

export default Card;
