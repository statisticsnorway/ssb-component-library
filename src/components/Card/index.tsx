import React, { type ReactNode } from 'react'
import { ArrowRight, Download, ExternalLink } from 'react-feather'

interface CardProps {
  ariaLabel?: string
  children?: ReactNode
  className?: string
  downloadText?: string
  external?: boolean
  fileLocation: string
  href: string
  hrefText?: string
  icon?: ReactNode
  image?: ReactNode
  imagePlacement?: 'left' | 'top'
  profiled?: boolean
  subTitle?: string
  title?: string
}

const Card: React.FC<CardProps> = ({
  ariaLabel,
  children,
  className,
  external,
  downloadText = 'Last ned',
  fileLocation,
  href,
  hrefText,
  icon,
  image,
  imagePlacement = 'top',
  profiled = false,
  subTitle,
  title,
}) => (
  <div className={`ssb-card${className ? ` ${className}` : ''}`}>
    <div
      className={`clickable ${imagePlacement === 'left' ? 'left-orientation' : 'top-orientation'}`}
      aria-label={ariaLabel}
    >
      {image && <div className='card-image'>{image}</div>}

      <div
        className={`card-content${image ? ' with-image' : ''}${profiled ? ' profiled' : ''}${external ? ' external' : ''}`}
      >
        {icon && <div className='card-icon'>{icon}</div>}
        {subTitle && <div className='card-subtitle'>{subTitle}</div>}
        {title && (
          <a
            href={href}
            className='card-title'
            target={external ? '_blank' : undefined}
            rel={external ? 'noreferrer' : undefined}
          >
            {title}
          </a>
        )}

        {children && <div className='card-text'>{children}</div>}

        {!image &&
          !hrefText &&
          (external ? (
            <ExternalLink className='arrow-icon' size={22} aria-hidden='true' />
          ) : (
            <ArrowRight className='arrow-icon' size={22} aria-hidden='true' />
          ))}

        {!title && !image && hrefText && (
          <a
            className='card-action'
            href={href}
            target={external ? '_blank' : undefined}
            rel={external ? 'noreferrer' : undefined}
          >
            {external ? (
              <ExternalLink className='arrow-icon' size={16} aria-hidden='true' />
            ) : (
              <ArrowRight className='arrow-icon' size={16} aria-hidden='true' />
            )}
            <div className='href-text'>{hrefText}</div>
          </a>
        )}

        {title && !image && hrefText && (
          <div className='card-action'>
            {external ? (
              <ExternalLink className='arrow-icon' size={16} aria-hidden='true' />
            ) : (
              <ArrowRight className='arrow-icon' size={16} aria-hidden='true' />
            )}
            <div className='href-text'>{hrefText}</div>
          </div>
        )}
      </div>
    </div>
    {fileLocation && (
      <a download href={fileLocation} className='download-section'>
        <Download className='download-icon' size={22} />
        <span>{downloadText}</span>
      </a>
    )}
  </div>
)

export default Card
