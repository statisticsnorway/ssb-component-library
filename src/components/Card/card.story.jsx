import React from 'react';
import { storiesOf } from '@storybook/react';
import centered from '@storybook/addon-centered/react';
import Card from './index';
import Button from '../Button';
import Paragraph from '../Paragraph';
import Title from '../Title';

storiesOf('Card', module).addDecorator(centered)
	.add('Default', () => (
		<div style={{ width: '300px' }}>
			<Card>
				<Title size={4} className="no-margin">Look at this</Title>
				<Paragraph>Explain something about something with something clever.</Paragraph>
				<Button primary>Handling</Button>
			</Card>
		</div>
	))
	.add('Three boxes', () => (
		<div style={{ display: 'grid', gridTemplateColumns: 'repeat(3, 250px)', gridColumnGap: '20px' }}>
			<Card>
				<Title size={2}>Look at this</Title>
				<Paragraph>Explain something about something with something clever.</Paragraph>
			</Card>
			<Card centered>
				<Title size={2}>Look at this</Title>
				<Paragraph>This box has the <code>centered</code> prop set to true</Paragraph>
			</Card>
			<Card fileLocation="./not_a_file.md">
				<Title size={2}>Look at this</Title>
				<Paragraph>Explain something about something with something clever.</Paragraph>
			</Card>
		</div>
	));
