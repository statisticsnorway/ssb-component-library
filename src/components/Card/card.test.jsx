import React from 'react';
import {shallow} from 'enzyme';
import Card from './index';

describe('Card component', () => {
	test('Matches the snapshot', () => {
		const wrapper = shallow(<Card>ProfileBox</Card>);
		expect(wrapper).toMatchSnapshot();
	});
	test('Toggles classNames centered correctly', () => {
		const wrapper = shallow(<Card centered>ProfileBox</Card>);
		expect(wrapper.find('.card-content').hasClass('centered')).toEqual(true);
	});
	test('Toggles classNames image correctly', () => {
		const wrapper = shallow(<Card image={<img src="testImage" alt="testImage" />}>ProfileBox</Card>);
		expect(wrapper.find('button').containsMatchingElement(<img/>)).toEqual(true);
	});
	test('Show image orientation correctly', () => {
		const wrapper = shallow(<Card imagePlacement="left">ProfileBox</Card>);
		expect(wrapper.find('button').hasClass('left-orientation')).toEqual(true);
	});
	test('Show download link if fileLocation', () => {
		const wrapper = shallow(<Card fileLocation="./not_a_file.md" downloadText="Download">ProfileBox</Card>);
		expect(wrapper.find('a').hasClass('download-section')).toEqual(true);
		expect(wrapper.containsMatchingElement(<span>Download</span>)).toEqual(true);
	});

});
