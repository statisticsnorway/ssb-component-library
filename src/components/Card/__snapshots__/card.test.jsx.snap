// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component Matches the snapshot 1`] = `
<div
  className="ssb-card"
>
  <button
    className="clickable top-orientation"
    onClick={[Function]}
  >
    <div
      className="card-content"
    >
      ProfileBox
    </div>
  </button>
</div>
`;
